File: src/recursion/subsets.ts
// DIFFICULTY: MEDIUM
//
// Given an integer array nums of unique elements, return all possible subsets  (the power set).
//
// The solution set must not contain duplicate subsets. Return the solution in any order.
//
// See {@link https://leetcode.com/problems/subsets/}
export { subsets };

// SOLUTION:
//
// To do this, recursively generate all combinations of elements of the set `xs - x`, then take all the subsets
// created and add x to them.  This should result in 2^n number of subsets.
function subsets(nums: number[]) {
  function run(xs: Set<number>): Set<Set<number>> {
    if (xs.size === 0) {
      // Note that new Set(new Set()) gives you a set with zero elements, because the constructor takes all elements of
      // the first argument and adds them to the set.  So you actually want to construct a new iterable based on the
      // empty set.
      return new Set([new Set()]);
    }

    // Find the first element of xs and call it x; remove it from the list.
    const x = xs.values().next().value!;

    // Generate all subsets without x.
    const rest = new Set(xs);
    rest.delete(x);

    const excluded: Set<Set<number>> = run(rest);

    // Add x to all of the subsets from above.
    const included = new Set<Set<number>>();
    excluded.forEach(ex => {
      const set = new Set(ex);
      set.add(x);

      included.add(set);
    });

    // The subsets with x and the subsets without x constitute the power set.
    return new Set([...excluded, ...included]);
  }

  // LeetCode expects the result to be an array of arrays.
  const result = run(new Set(nums));
  return [...result].map(xs => [...xs]);
}

File: src/leetcode/recursion/subsets.py
# DIFFICULTY: MEDIUM
# ------------------
#
# Given an integer array nums of unique elements, return all possible subsets  (the power set).
#
# The solution set must not contain duplicate subsets. Return the solution in any order.
#
# See https://leetcode.com/problems/subsets
class Solution:
    def subsets(self, nums: list[int]) -> list[list[int]]:
        """
        SOLUTION
        --------
        To do this, recursively generate all combinations of elements of the set `xs - x`, then take all the subsets
        created and add x to them.  This should result in 2^n number of subsets.

        COMPLEXITY
        ----------
        Time complexity is O(2^n) where n is the number of elements in the list.

        Space complexity is O(n) because we are storing the subsets in stack frames.
        """

        # Note that you cannot put a set inside of a set; sets are mutable and cannot be hashed.  However, frozensets
        # are immutable, so you can use those instead.
        def generate(xs: set[int]) -> set[frozenset[int]]:
            if not xs:
                # Note that set(frozenset()) creates an empty set, not a set with an empty set in it.
                return {frozenset()}

            # Find the first element of xs and call it x; remove it from the set and call it rest.
            ys = list(xs)
            x = ys[0]
            rest = set(ys[1:])

            # Generate all subsets of that set.
            excludeds = generate(rest)

            # Now generate all subsets with x by adding x to each excluded subset.
            includeds = {subset | {x} for subset in excludeds}

            # Combine the excluded and included subsets.
            return excludeds | includeds

        result = generate(set(nums))
        return [list(subset) for subset in result]
